content/academy/caching_responses_in_puppeteer.md

```diff
@@ -0,0 +1,169 @@
+        }
+
+        cache[url] = {
+            status: response.status(),
+            headers: response.headers(),
+            body: buffer,
+            expires: Date.now() + (maxAge * 1000),
+        };
+    }
+});
+```
+
+> If the code above looks completely foreign to you, we recommending going through our free [Puppeteer/Playwright course]({{@link puppeteer_playwright.md}}).
+
+After implementing this code, we can run the scraper again.
+
+![Good run results]({{@asset images/good-run-results.webp}})
+
+Looking at the statistics, caching responses in Puppeteer brought the traffic down from 177MB to 13.4MB, which is a reduction of data transfer by 92%. The related screenshots can be found [here](https://my.apify.com/storage/key-value/iWQ3mQE2XsLA2eErL).
+
+It did not speed up the crawler, but that is only because the crawler is set to wait until the network is nearly idle, and CNN has a lot of tracking and analytics scripts that keep the network busy.
+
+## Implementation in Crawlee
+
+Since most of you are likely using [Crawlee](https://crawlee.dev), here is what response caching would look like using `PuppeteerCrawler`:
+
+```JavaScript
+import { PuppeteerCrawler, Dataset } from 'crawlee';
+
+const cache = {};
+
+const crawler = new PuppeteerCrawler({
+    preNavigationHooks: [async ({ page }) => {
+        await page.setRequestInterception(true);
+
+        page.on('request', async (request) => {
+            const url = request.url();
+            if (cache[url] && cache[url].expires > Date.now()) {
+                await request.respond(cache[url]);
+                return;
+            }
+            request.continue();
+        });
+
+        page.on('response', async (response) => {
+            const url = response.url();
+            const headers = response.headers();
+            const cacheControl = headers['cache-control'] || '';
+            const maxAgeMatch = cacheControl.match(/max-age=(\d+)/);
+            const maxAge = maxAgeMatch && maxAgeMatch.length > 1 ? parseInt(maxAgeMatch[1], 10) : 0;
+
+            if (maxAge) {
+                if (!cache[url] || cache[url].expires > Date.now()) return;
+
+                let buffer;
+                try {
+                    buffer = await response.buffer();
+                } catch (error) {
+                    // some responses do not contain buffer and do not need to be catched
+                    return;
+                }
+
+                cache[url] = {
+                    status: response.status(),
+                    headers: response.headers(),
+                    body: buffer,
+                    expires: Date.now() + maxAge * 1000,
+                };
+            }
+        });
+    }],
+    requestHandler: async ({ page, request }) => {
+        await Dataset.pushData({
+            title: await page.title(),
+            url: request.url,
+            succeeded: true,
+        });
+    },
+
+    failedRequestHandler: async ({ request }) => {
```

Comment:
Let's remove this, there is an agreement now that pushing errors into data in anti-pattern